### ISSUE TITLE
Aggregate / sum columns

### ISSUE THREAD

**Issue by dkoo761** [aldeed/meteor-tabular#101]
I have a table of users where, for each user, I'd like to:

1) include 3 columns, with each column showing an aggregate/sum of the number of times the user has performed some action. I have a separate "actions" collection with 1 document per user action.

2) allow the table to be sorted by each of the aggregate action columns

3) allow the user to use a 'start date' and 'end date' datepicker to filter the action count columns to only include an aggregation of actions that were performed within the date range.

So something like this:

```
| Name       | Action 1 | Action 2 | Action 3 |
| John Smith | 13       | 25       | 2        |
| Jane Smith | 23       | 15       | 5        |
```

I've narrowed down to 2 possible approaches:

1) Compute aggregations on the client

I could use a collection helper, render() function or template for each action column and deal with the performance hit of doing it client-side. In this case, I would use publishComposite to return all the actions for each user.

PROS: No need to create a collection on the server with transient document set per client-request.

CONS: I would lose the ability to sort and would also have to apply the date range filters client-side, meaning that I would potentially get far more data in the "actions" subscription than would be actually needed to compute the aggregations. Performance would probably suffer as even with pagination, there could be hundreds of action documents published to the client per user.

2) Compute aggregations on the server and store them in transient documents

I could make a method call with the client's date range filters to tell the server to compute aggregations (with https://github.com/meteorhacks/meteor-aggregate) and save them as fields in transient documents in a "userActionCounts" collection on the server. Each document would correspond exactly to the data needed in 1 row of the table on the client (with 1 extra field to store the _id of the user performing the query). It would then be trivial to publish this collection and display the data using Tabular.

PROS: All requirements would be met (including sorting). Performance should be better.

CONS: I would have to maintain an extra userActionCounts collection on the server with these transient documents that only exist until the user leaves the page at which point the documents would be deleted from the collection. Also, each time the user updates their search params (date range), I would need to wipe out all their userActionCounts documents and insert a new set of documents with the new counts into the userActionCounts collection. We have some clients with over 1,000 employees so I would literally be computing and inserting 1,000 documents into the userActionCounts collection each time someone visited the page, ouch.

I'm thinking option 2 is probably the way to go. But would be interested to hear your opinion.

Thanks in advance!
Dave

**Comment by rhyslbw** [comment]
Would a non-reactive aggregation + polling approach work for your requirements? That would be most efficient and removes the need for data duplication just to fit it into this solution.

**Comment by dkoo761** [comment]
@rhyslbw Could you elaborate a bit more? I'm not quite sure I know what you mean. I think I would lose the ability to sort on the "count" columns with any solution that didn't involve a collection of transient documents that included count properties in each document.

**Comment by rhyslbw** [comment]
I'm just throwing some ideas out, but could you poll a Meteor method that calls the aggregate, then use Minimongo as a local cache?

**Comment by aldeed** [comment]
Typically this would be implemented with a special aggregation publication and a client-only collection (somewhat like the "counts-by-room" example [here](http://docs.meteor.com/#/full/publishandsubscribe).

I think this would require a lot of design work and some special support in this package to make it work smoothly. In the end it might be easier to store them in a separate collection, and then set up a cron job every 5 minutes or so to calculate and update the counts.

**Comment by dkoo761** [comment]
@rhyslbw Sure, if I didn't want to use Tabular then I could use a Meteor method with Mongo aggregations instead of pub/sub but I would prefer to stick with Tabular since I get search/sort/paginate  and DataTables integration for free.

@aldeed Thanks for the feedback. In my case, a cron wouldn't work since the counts in each row of the table are depending on the start/end date chosen by the user. So for now I've gone with my option #2 above and added a transient documents collection that gets populated each time the user chooses to view the report. Works fine for now and I can always refactor if the app reaches a scale where this solution is outgrown. At this stage, it's an MVP so we just need it working, not necessarily working fast ;)

**Comment by dkoo761** [comment]
A quick footnote in case anyone else is doing something like this. The following combination works really well for my Option 2 above:

- Tabular
- meteorhacks:aggregate
- collection of transient "report row" documents
- bulk insert of those transient documents using Node's lower-level Mongo driver (made available in Meteor 1.0.4 as ```rawDatabase``` and ```rawCollection```) as follows:

```javascript
var collection = SomeCollection.rawCollection();
var dbInsertSync = Meteor.wrapAsync(collection.insert, collection);
var batches = _.groupBy(reportRows, function (element, index) {
    return Math.floor(index / Constants.DB_INSERT_BATCH_SIZE);
});
_.each(batches, function (batch) {
    result = dbInsertSync(batch, null);
});
````

The reason for batching the inserts (I use a batch size of 1,000 documents) is to ensure that each batch (array) of documents doesn't exceed the 16MB Mongo limit.

Performance-wise, even if the report has thousands of rows, the counts are computed via a single DB aggregate query. And by combining that with bulk DB inserts, the result is sub-second real-time report generation with all the great Tabular features intact.

**Comment by rhyslbw** [comment]
:+1:
